### PR TITLE
Code cleanup

### PR DESCRIPTION
**Pull request by JezSonic**
Improve code quality and add linter checks to GIthub Actions

### CODE REVIEW

File: src/core/TrainController.hpp [hunk 1]
```diff
@@ -71,50 +72,53 @@ namespace godot {
 
             Dictionary get_config() const;
             void update_config(const Dictionary &p_config);
-            void set_config_property(String &key, Variant &p_value);
+            static void set_config_property(String &key, Variant &p_value);
```
Comment (marcinn) [inline src/core/TrainController.hpp:75]: ???

File: src/core/TrainPart.hpp [hunk 2]
```diff
@@ -53,24 +54,25 @@ namespace godot {
             virtual void _register_commands();
             virtual void _unregister_commands();
 
-            TMoverParameters *get_mover();
+            [[nodiscard]] TMoverParameters *get_mover() const;
```
Comment (marcinn) [inline src/core/TrainPart.hpp:57]: ???

File: src/core/TrainSystem.cpp [hunk 3]
```diff
@@ -77,19 +78,20 @@ namespace godot {
...
     Dictionary TrainSystem::get_all_config_properties(const String &train_id) {
-        auto it = trains.find(train_id);
+        const std::map<String, TrainController *>::iterator it = trains.find(train_id);
 
         if (it == trains.end()) {
-            log(train_id, LogSystem::LogLevel::LOGLEVEL_ERROR, "Train is not registered in");
-            UtilityFunctions::push_error("Train is not registered: ", train_id);
-            return Dictionary();
+            log(train_id, LogSystem::LogLevel::ERROR, "Train is not registered in");
+            UtilityFunctions::push_error("Train is no{}: ", train_id);
```
Comment (marcinn) [inline src/core/TrainSystem.cpp:85]: Why?

**Review by marcinn — CHANGES_REQUESTED** [review]
Too many changes in one PR :(

Please split to a separate PR with formatting/linters/c++ compat changes (please also add GH action with formatting checks). Please avoid linter exceptions in comments - it's better to configure the tool correctly and have clean code. Also, as we talked on the Discord server, I see inconsistencies in the use of consts and some API changes that should be discussed. 

It would be really better to do this on a separate branch. For now we need to quickfix LogSystem/demo issues. Once that's done, I'll do another code review, this time in the context of the LogSystem fixes. 

Thanks!

File: src/core/TrainPart.hpp (hunk 2 above)
Comment (JezSonic) [reply src/core/TrainPart.hpp:56]: https://stackoverflow.com/questions/76489630/explanation-of-nodiscard-in-c17

File: src/core/TrainController.hpp (hunk 1 above)
Comment (JezSonic) [reply src/core/TrainController.hpp:74]: It was possible to make it static without crashing anything, so I did that

**Comment by marcinn** [comment]
Ok, I think you're on the right track, but there are a few obstacles not related to clang-tidy itself:
* setup `.clang-tidy` to exactly check what we need
* match `.clang-format` with  `.clang-tidy` rules (this is for autoformatting the code, which can be enabled in IDEs/Editors)
* run `scons compiledb` to generate `compile_commands.json` before runing the clang-tidy (it **must** be run in the docker after checking out the sources and before running clang-tidy)
* setup proper include and exclude paths

### Issues with include / exclude

Please note that you're **wrapping** clang-tidy with a custom action `clang-tidy-review`, which is built top of a Docker (! another one? Docker in docker?) and wraps `clang-tidy` with custom python scripts.  It has `include:` and `exclude:` options, but the `cwd` (current working directory) could be different than we assume.  Try to prefix patterns with absolute paths to the checked out sources. 

Also please note, that action you used is designed to run only for changes. I don't know what it means, but maybe it just runs once per commit?

image

It is also possible that the action you used may be faulty, and you're struggling not with clang, but someone's wrapper.  Please consider using a more typical `clang-tidy` action, which will also check all sources. I think that we don't need to check only changes, because all our sources should follow coding standards (excl. Mover sources and docgen).

Alternatively you can just run this command on checked out sources (I assume that clang-tidy and scons are installed, and cwd is set to the repository directory):

```
scons compiledb && \
   find src/ \( -name "*.cpp" -or -name "*.hpp" \) \
	-not -path "src/maszyna/*"\
	-not -path "src/gen/*"\
	-exec clang-tidy --fix-notes --quiet -p . {} +
```

This command will generate `compile_commands.json`, then will find sources (`*.hpp` and `*.src`) in `src/`, but excluding `src/maszyna` (legacy code) and `src/gen` (from doc generator).

----

As PR states about cleaning the code, the final step is to run clang-tidy with `--fix-errors` option, **and** fix all issues. We can do this together. But because we have other PR's opened and fixes could generate quite big diffs and many conflicts, I suggest to merge existing PRs first (same as with crlf).

Here is proposal of my `.clang-tidy` and `.clang-format`, together with clang-tidy invocations as Makefile targets: https://github.com/MaSzyna-Reloaded/MaSzyna-API-wrapper/pull/77/files
I also tried to add GitHub Workflow action, but it founds no files to check (as I wrote above).

Fingers crossed!  :crossed_fingers:

**Comment by marcinn** [comment]
Oh, and one more thing. It's fine for me to also add Qodana checks, but only if their results match clang-tidy. The goal is to have consistent feedback from our tools.

**Comment by marcinn** [comment]
I think these comments are unnecessary. Successful builds on PR are already marked.

image

File: src/core/TrainController.cpp [hunk 4]
```diff
@@ -106,9 +105,8 @@ namespace godot {
                 "get_battery_voltage");
     }
 
-    TrainController::TrainController() {
+    TrainController::TrainController() = default;
```
[the remark this answers is not in the document]
Comment (carmel4a) [reply src/core/TrainController.cpp:108]: move ctr into `hpp`

**Review by carmel4a — CHANGES_REQUESTED** [review]
Review still in progress, main concerns:
- const's in setters
- issue with default constructors -> move to hpp
-

File: .github/workflows/qodana.yaml [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        with:
+          submodules: recursive
+          ref: ${{ github.event.pull_request.head.sha }}  # to check out the actual pull request commit, not the merge commit
+          fetch-depth: 0  # a full history is required for pull request analysis
+      - name: 'Qodana Scan'
+        uses: JetBrains/qodana-action@v2024.2.3
+        with:
+          pr-mode: false
+        env:
+          QODANA_TOKEN: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwcm9qZWN0IjoiQVJqSjYiLCJvcmdhbml6YXRpb24iOiJBYUc3RCIsInRva2VuIjoicDJNWVcifQ.myd_bOxtnHr9gwQr7d6-FlvO7n_T6RwIH20H40XYTQ0'
```
[the remark this answers is not in the document]
Comment (JezSonic) [reply .github/workflows/qodana.yaml:28]: This might lead to fails on forks but I think there's a workaround for that

File: CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
+option(QODANA_BUILD "If the build is for Qodana CI/CD" OFF)
+
+AUX_SOURCE_DIRECTORY(src SRCES)
+if (NOT QODANA_BUILD) #Qodana will run inspection on godot-cpp too. We don't need that
+    message("Not building for Qodana")
+    add_subdirectory(godot-cpp)
+else ()
+    message("Building for Qodana")
+endif()
```
[the remark this answers is not in the document]
Comment (JezSonic) [reply CMakeLists.txt:20]: Cause (for both comments) we will get code scanning results related to godot-cpp's code and we don't want that

File: src/brakes/TrainBrake.cpp [hunk 7]
```diff
@@ -238,51 +237,51 @@ namespace godot {
...
 
-    void TrainBrake::brake_releaser(const bool p_pressed) {
+    void TrainBrake::brake_releaser(const bool p_pressed) const {
         TMoverParameters *mover = get_mover();
         ASSERT_MOVER_BRAKE(mover);
         mover->BrakeReleaser(p_pressed ? 1 : 0);
     }
 
-    void TrainBrake::brake_level_set(const float p_level) {
+    void TrainBrake::brake_level_set(const float p_level) const {
```
[the remark this answers is not in the document]
Comment (JezSonic) [reply src/brakes/TrainBrake.cpp:246]: Move this into discussion please, for now we can leave it i think

File: demo/debug_panel.tscn [hunk 8]
```diff
@@ -16,9 +16,9 @@ border_width_bottom = 1
 border_color = Color(0.160034, 0.223886, 0.359375, 1)
 
 [node name="DebugPanel" type="VBoxContainer"]
-custom_minimum_size = Vector2(0, 400)
+custom_minimum_size = Vector2(0, 250)
 offset_right = 258.0
-offset_bottom = 369.0
+offset_bottom = 400.0
```
Comment (marcinn) [inline demo/debug_panel.tscn:21]: is this change related to the code cleanup?

File: demo/debug_panel.tscn [hunk 9]
```diff
@@ -16,9 +16,9 @@ border_width_bottom = 1
 border_color = Color(0.160034, 0.223886, 0.359375, 1)
 
 [node name="DebugPanel" type="VBoxContainer"]
-custom_minimum_size = Vector2(0, 400)
+custom_minimum_size = Vector2(0, 250)
```
Comment (marcinn) [inline demo/debug_panel.tscn:19]: is this change related to the code cleanup?

File: demo/demo.tscn [hunk 10]
```diff
@@ -25,10 +25,10 @@ script = ExtResource("1_76ghk")
 
 [node name="UI" type="VBoxContainer" parent="."]
 layout_mode = 1
-anchors_preset = -1
+anchors_preset = 15
 anchor_right = 1.0
-anchor_bottom = 1.02
-offset_bottom = 0.039917
+anchor_bottom = 1.0
+offset_bottom = -3.0
```
Comment (marcinn) [inline demo/demo.tscn:31]: is this change related to the code cleanup? (same for above changes)

File: src/core/TrainController.cpp [hunk 11]
```diff
@@ -431,6 +433,8 @@ namespace godot {
         emit_signal(CONFIG_CHANGED);
     }
 
+    void TrainController::set_config_property(String &key, Variant &p_value) {}
```
Comment (marcinn) [inline src/core/TrainController.cpp:436]: Why? Is this change related to the code cleanup? The config property is READ-ONLY.

File: src/core/TrainController.cpp [hunk 12]
```diff
@@ -339,14 +339,16 @@ namespace godot {
...
-    void TrainController::set_axle_arrangement(String p_value) {
+    void TrainController::set_axle_arrangement(const String &p_value) {
         axle_arrangement = p_value;
     }
 
     String TrainController::get_axle_arrangement() const {
         return axle_arrangement;
     }
 
+    void TrainController::set_state(const Dictionary &p_state) {}
```
Comment (marcinn) [inline src/core/TrainController.cpp:350]: Why? Is this change related to the code cleanup? The state property is READ-ONLY.

File: src/core/TrainController.hpp [hunk 13]
```diff
@@ -33,15 +33,15 @@ namespace godot {
             double max_velocity = 0.0;
             int radio_channel = 0;
             int radio_channel_min = 0;
-            int radio_channel_max = 10;
+            int radio_channel_max = 10; //TODO: Move these to consts to avoid "magic numbers"?
```
Comment (marcinn) [inline src/core/TrainController.hpp:36]: It's a configuration value.
Radio will be cleaned and moved to a separate node in #33

File: src/core/TrainController.hpp [hunk 14]
```diff
@@ -71,54 +71,54 @@ namespace godot {
 
             Dictionary get_config() const;
             void update_config(const Dictionary &p_config);
-            void set_config_property(String &key, Variant &p_value);
+            static void set_config_property(String &key, Variant &p_value);
             void _process(double delta) override;
+            // ReSharper disable once CppHidingFunction
```
Comment (marcinn) [inline src/core/TrainController.hpp:76]: Still there is a comment for a specific linter. How about others?

File: src/core/TrainPart.hpp [hunk 15]
```diff
@@ -21,6 +20,7 @@ namespace godot {
             bool _commands_registered = false;
 
         protected:
+            // ReSharper disable once CppHidingFunction
```
Comment (marcinn) [inline src/core/TrainPart.hpp:23]: specific linter configuration. how about others?

File: src/core/TrainPart.hpp (hunk 2 above)
Comment (marcinn) [reply src/core/TrainPart.hpp:56]: I do not understand. How about other methods? They have return values, too, but without `[[nodiscard]]`.
image

`get_mover()` is not changing anything. I do not understand.

File: src/core/TrainSystem.cpp (hunk 3 above)
Comment (marcinn) [reply src/core/TrainSystem.cpp:83]: still not fixed

**Review by marcinn — CHANGES_REQUESTED** [review]
Please remove changes not related to the cleaning up.
Please correct changes made by mistake.

Consider disabling github-bot or adapting it to our needs. It should NOT check `src/machine` or suggest/fix code according to rules not adopted by us (see `[[nodiscard]]` and others to discuss)

File: demo/debug_panel.tscn (hunk 8 above)
Comment (JezSonic) [reply demo/debug_panel.tscn:21]: mistake, sorry, but had top do that in order to display things properly on my screen

**Comment by JezSonic** [comment]
> Please remove changes not related to the cleaning up. Please correct changes made by mistake.
> 
> Consider disabling github-bot or adapting it to our needs. It should NOT check `src/machine` or suggest/fix code according to rules not adopted by us (see `[[nodiscard]]` and others to discuss)

The thing is that we haven't adapted them properly (we did code according to older standards than C++17) so there's this code cleanup for. Comments from old action are disabled but Qodana will still comment on PRs.

For now `[[nodiscard]]` is only new directive - we should enforce using that
We should also not declare constructors and destructors at all unless we need them for specific behavior. C++is made for compilers to generate them automatically if they are not declared.

It WILL check `source/maszyna` code cause it is our code, not propertiary library like `godot-cpp`. Besides we still have our custom code in that folder. I see no point in disabling this and rather I'd be against excluding this directory

File: src/core/TrainPart.hpp (hunk 15 above)
Comment (JezSonic) [reply src/core/TrainPart.hpp:23]: Clang-Tidy have not reported anything as well. This might be old linter configuration (before adding your .clang-tidy). which I have forgot to remove I'll see if it is still relevant there or not.

File: src/core/TrainController.cpp (hunk 12 above)
Comment (JezSonic) [reply src/core/TrainController.cpp:350]: Method was declared in .hpp file but not in .cpp

So i just have declared it in .cpp

File: src/core/TrainController.cpp (hunk 11 above)
Comment (JezSonic) [reply src/core/TrainController.cpp:436]: Method was declared in .hpp file but not in .cpp

So i just have declared it in .cpp

**Comment by marcinn** [comment]
It must not check Mover's code. It will generate mess comments/suggestions otherwise. Clang tidy config I suggested earlier is doing that exclusion. 

We will never touch Mover's code to match these rules.

----

TBH I never saw nodiscard directive. And I see that it is not applied everywhere, so probably there is an inconsistency. 

All of us must know when to use it and what for it is, and our all tools should inform us about lack of it. But I never saw such suggestion from clang-tidy.

File: src/core/TrainController.hpp [hunk 16]
```diff
@@ -41,9 +41,7 @@ namespace godot {
 
             String axle_arrangement = "";
 
-            void _collect_train_parts(const Node *node, Vector<TrainPart *> &train_parts);
-
-        private:
+            void _collect_train_parts(const Node *node, Vector<TrainPart *> &train_parts) {};
```
Comment (marcinn) [inline src/core/TrainController.hpp:44]: As far as I can see this is an unused method. Just delete it.

File: src/core/TrainController.hpp [hunk 17]
```diff
@@ -71,19 +69,19 @@ namespace godot {
 
             Dictionary get_config() const;
             void update_config(const Dictionary &p_config);
-            void set_config_property(String &key, Variant &p_value);
+            void set_config_property(String &key, Variant &p_value) {};
```
Comment (marcinn) [inline src/core/TrainController.hpp:72]: This should not be implemented as empty method. This is unused method, so declaration can be safely removed.

File: src/core/TrainController.hpp [hunk 18]
```diff
@@ -99,26 +97,23 @@ namespace godot {
...
             int get_radio_channel_min() const;
-            void set_radio_channel_min(const int p_value);
+            void set_radio_channel_min(int p_value);
             int get_radio_channel_max() const;
-            void set_radio_channel_max(const int p_value);
+            void set_radio_channel_max(int p_value);
             String get_axle_arrangement() const;
 
-            void set_state(const Dictionary &p_state);
+            void set_state(const Dictionary &p_state) {};
```
Comment (marcinn) [inline src/core/TrainController.hpp:115]: This should not be implemented as empty method. This is unused method, so declaration can be safely removed.

File: src/core/TrainPart.cpp [hunk 19]
```diff
@@ -80,6 +79,9 @@ namespace godot {
                 }
                 train_controller_node = nullptr;
             } break;
+            default:
+                DEBUG("TrainController::_ready() This notification type is not implemented");
```
Comment (marcinn) [inline src/core/TrainPart.cpp:83]: There should be no warning message, because it will generate spam messages - not all notifications must be handled.

File: src/core/TrainSystem.cpp [hunk 20]
```diff
@@ -77,19 +71,19 @@ namespace godot {
...
 
     Dictionary TrainSystem::get_all_config_properties(const String &train_id) {
-        auto it = trains.find(train_id);
+        const std::map<String, TrainController *>::iterator it = trains.find(train_id);
 
         if (it == trains.end()) {
             log(train_id, LogSystem::LogLevel::ERROR, "Train is not registered in");
             UtilityFunctions::push_error("Train is not registered: ", train_id);
-            return Dictionary();
+            return {};
```
Comment (marcinn) [inline src/core/TrainSystem.cpp:79]: AFAIK `{}` is not a `Dictionary()` ?

File: src/core/TrainSystem.hpp [hunk 21]
```diff
@@ -1,19 +1,18 @@
...
 #include <godot_cpp/variant/utility_functions.hpp>
 #include <map>
-#include "./LogSystem.hpp"
 
 namespace godot {
 
     class TrainController;
 
-    class TrainSystem : public RefCounted {
+    class TrainSystem final : public RefCounted {
```
Comment (marcinn) [inline src/core/TrainSystem.hpp:15]: I don't see the point of "final" here. It could be added, of course, because it changes nothing for now.

File: src/core/TrainSystem.hpp [hunk 22]
```diff
@@ -23,20 +22,17 @@ namespace godot {
         public:
             static const char *TRAIN_LOG_UPDATED_SIGNAL;
 
-            inline static TrainSystem *get_instance() {
+            static TrainSystem *get_instance() {
```
Comment (marcinn) [inline src/core/TrainSystem.hpp:25]: Why not inline?

File: src/doors/TrainDoors.cpp [hunk 23]
```diff
@@ -215,12 +214,12 @@ namespace godot {
         state["doors_right_remote_open"] = right_door.remote_open;
         state["doors_right_position"] = right_door.position;
         state["doors_right_position_normalized"] = right_door.position / max_shift;
-        state["doors_right_operating"] = right_door.is_opening | right_door.is_closing;
+        state["doors_right_operating"] = static_cast<int>(right_door.is_opening) | static_cast<int>(right_door.is_closing);
```
Comment (marcinn) [inline src/doors/TrainDoors.cpp:217]: changes bool to int. should be bool.

File: src/doors/TrainDoors.hpp [hunk 24]
```diff
@@ -161,22 +158,22 @@ namespace godot {
             /**
              * Time period for which door would stay opened. Default value from internal mover
              */
-            float open_time = -1.0f;
+            float open_time = -1.0F;
```
Comment (marcinn) [inline src/doors/TrainDoors.hpp:161]: why F not f?

File: src/doors/TrainDoors.hpp [hunk 25]
```diff
@@ -249,13 +246,13 @@ namespace godot {
...
              */
-            float platform_max_shift = 0.0f;
+            float platform_max_shift = 0.0F;
 
             /**
              * The speed of the animation step, where 1.0 corresponds to an animation lasting one second, a value of 0.5
              * to two seconds, etc. Default value from internal mover
              */
-            float platform_speed = 0.5f;
+            float platform_speed = 0.5F;
```
Comment (marcinn) [inline src/doors/TrainDoors.hpp:255]: same as above... why F not f.

File: src/editor/DieselEngineMasterControllerPowerItemEditor.hpp [hunk 26]
```diff
@@ -11,7 +10,8 @@
 
 using namespace godot;
 
-class DieselEngineMasterControllerPowerItemEditor : public EditorProperty {
+class DieselEngineMasterControllerPowerItemEditor final : public EditorProperty {
```
[the remark this answers is not in the document]
Comment (marcinn) [reply src/editor/DieselEngineMasterControllerPowerItemEditor.hpp:13]: please remove DieselEngineMasterControllerPowerItemEditor - it is unused

File: src/engines/TrainDieselElectricEngine.hpp [hunk 27]
```diff
@@ -1,17 +1,14 @@
+#pragma once
 #include "TrainDieselEngine.hpp"
 
 namespace godot {
-    class TrainDieselElectricEngine : public TrainDieselEngine {
+    class TrainDieselElectricEngine final : public TrainDieselEngine {
```
Comment (marcinn) [inline src/engines/TrainDieselElectricEngine.hpp:5]: "Final" shouldn't be here

File: src/engines/TrainEngine.cpp [hunk 28]
```diff
@@ -82,9 +78,9 @@ namespace godot {
         return motor_param_table;
     }
 
-    void TrainEngine::set_motor_param_table(const TypedArray<Dictionary> p_value) {
+    void TrainEngine::set_motor_param_table(const TypedArray<Dictionary> &p_wwlist) {
```
Comment (marcinn) [inline src/engines/TrainEngine.cpp:81]: motor_param_table is not a wwlist

File: src/engines/TrainEngine.hpp [hunk 29]
```diff
@@ -22,10 +22,7 @@ namespace godot {
 
         public:
             TypedArray<Dictionary> get_motor_param_table();
-            void set_motor_param_table(const TypedArray<Dictionary> p_wwlist);
-            void main_switch(const bool p_enabled);
-
-            TrainEngine();
-            ~TrainEngine() override = default;
+            void set_motor_param_table(const TypedArray<Dictionary> &p_wwlist);
```
Comment (marcinn) [inline src/engines/TrainEngine.hpp:25]: motor_param_table is not a wwlist

File: src/plugin/LibMaszynaPlugin.hpp [hunk 30]
```diff
@@ -5,25 +5,25 @@
 
 #include <godot_cpp/classes/editor_plugin.hpp>
 #include <godot_cpp/core/class_db.hpp>
-#include <godot_cpp/godot.hpp>
 
 using namespace godot;
 
-class LibMaszynaPlugin : public EditorPlugin {
+class LibMaszynaPlugin final : public EditorPlugin {
```
[the remark this answers is not in the document]
Comment (marcinn) [reply src/plugin/LibMaszynaPlugin.hpp:11]: LibMaszynaPlugin is unused. Could be removed.

File: src/systems/TrainSecuritySystem.hpp [hunk 31]
```diff
@@ -1,11 +1,10 @@
 #pragma once
-#include <godot_cpp/classes/node.hpp>
 #include "../core/TrainPart.hpp"
 #include "../maszyna/McZapkie/MOVER.h"
-
+#include <godot_cpp/classes/node.hpp>
 
 namespace godot {
-    class TrainSecuritySystem : public TrainPart {
+    class TrainSecuritySystem final : public TrainPart {
```
Comment (marcinn) [inline src/systems/TrainSecuritySystem.hpp:7]: Why final here?

**Review by marcinn — CHANGES_REQUESTED** [review]

**Comment by marcinn** [comment]
One more thing:
image
Clang-tidy reports errors. Is the code not fully cleaned up or is there something wrong with the rules/configuration?

File: .gitignore [hunk 32]
```diff
@@ -27,4 +28,13 @@ docker-storage/MacOSX14.5.sdk.tar.xz
 /modules
 *.macos
 CMakeFiles/
-compile_commands.json
+/demo/e3d-test
```
Comment (marcinn) [inline .gitignore:31]: not related

File: src/doors/TrainDoors.cpp (hunk 23 above)
Comment (JezSonic) [reply src/doors/TrainDoors.cpp:217]: I took more brief look into that

```cpp
right_door.is_opening | right_door.is_closing;
```

`|` Implies conversion `bool -> init` as `|` is not an `OR` operator. `||` is, so I presume `|| should be there`

Thanks for pointing this out
